docs/design/schedule-customisation-design.md

```diff
@@ -0,0 +1,29 @@
+* As of now, users can specify actions and plugins through configmap, providing the same scheduling algorithms across the cluster. The aim of the project is to provide options to use different scheduling algorithms within the same cluster also. Users should be able to provide a different scheduling configuration for each Queue.
+
+## Solution
+We could either take actions and plugins directly in Queue CRD spec or we could have different config maps corresponding to different queues and take a configmap selector in Queue spec.  The current scenario which is cluster-wise configmap can be the default case when no particular scheduling config is specified with a queue. 
+
+Currently, the scheduling configuration is loaded at the beginning, and it's only updated when the underlying scheduling config is changed. This needs to be modified to load scheduling configuration before each session. Throughout a scheduling session, we can't change the configuration in between. Hence configuration has to be loaded before a scheduling session and shouldn't be changed till that session is ended.
+Currently, in a scheduling session, the scheduler traverses through actions ( which were loaded from the scheduling config in the beginning ). In each action, it goes through all jobs ( from all Queues ) which are yet to be scheduled and performs actions on them. If we want different Queues to have different configurations, then this cannot happen. 
+
+The scheduler should store different scheduling configurations and corresponding queues. After every 'SchedulePeriod', the scheduler can go through each configuration and create a separate session for it. For each session, it will only consider queues with that configuration. Action should take an input of a list of queues. So in each session, the scheduler will traverse through actions defined in the corresponding configmap. For each action, it should pass a list of queues. Each action executed should only consider jobs belonging to that list of queues.
+
+![Feature Design](./images/reservation_design.png)
+
+## Implementation
+1. Update the CRD structure to take scheduling configmap key in QueueSpec
+		The optional field which will take configmap selector
+		Default configmap is passed as a command-line argument to the scheduler it should be considered for a Queue in which no scheduling configmap is specified.
+2. Load and store different configurations in the scheduler.
+3. Store mapping of configuration to list of Queues in the scheduler. When a new Queue is created, its controller should check if a config map is specified with it. The controller should signal the scheduler to update its mapping of config to a list of queues to include a new queue in the corresponding list. 
+4. Traverse through each configuration periodically and create separate scheduling sessions for them.
```

Comment:
Are there any restrictions for doing it in one session?

Reply:
IMO, if we start different sessions for different configured queues, it is difficult to expand, what if we want to custom scheduling algorithms for per ns or maybe other custom requirements?

Reply:
> IMO, if we start different sessions for different configured queues, it is difficult to expand, what if we want to custom scheduling algorithms for per ns or maybe have other customed requirements later?

I think custom scheduling algorithms per ns is not so reasonable, but reasonable per queue.

Reply:
The biggest difficulity is to satisfy the requirement without changing existing use experience. It must keep backward capability.

Reply:
@shinytang6 
> Are there any restrictions for doing it in one session?

So a session is traversal of actions, now which actions are invoked depends on configuration right?
Different scheduling configurations will have different actions. So a session should just follow actions & plugins from one configuration. Once a config is loaded and session is started, it shouldn't change config until that session is closed. Hence we need separate sessions for separate configurations, right?

Reply:
There are two options :

1) We follow default config ( loaded through command-line ) in all cases except for custom queues. For custom queues, we could have config within yaml. So we load config and create a separate session for them. So in every iteration, we create one session for all the queues in default bucket, and a separate session per custom queue. 
2) We have a mapping of configuration to list of queues. By default, all queues will be mapped to default config ( loaded through command-line ). Now for custom queues, each custom queue will map to a config. Now in every iteration, scheduler will create a session for each configuration, it will handle all queues belonging to a configuration together.

Suppose we have a case where we have 200 queues. 100 of them follow default config. So there are 100 custom queues. Now in these 100 custom queues, if we have only 4 unique configurations ( say 25 queues per config ). Then
In design 1) we will have one session for default queues and 100 separate sessions for each custom queue. 
But in design 2) we will have only 5 sessions ( 1 for default config with 100, and 4 for custom config with 25 queues each ).